File: tools/glass-pane/backend/models/job.py
"""
Job-related data models.
"""

from pydantic import BaseModel
from typing import Optional, List, Dict, Any
from .phase import Phase, PhaseInfo


class Job(BaseModel):
    """Represents a Context Foundry build job."""

    id: str
    type: str  # 'autonomous_build', etc.
    status: str  # 'running', 'completed', 'failed', 'queued', 'cancelled'
    started_at: Optional[str] = None  # ISO 8601 timestamp
    completed_at: Optional[str] = None
    created_at: str  # ISO 8601 timestamp
    project_name: Optional[str] = None
    current_phase: Optional[Phase] = None
    tokens_used: int = 0
    total_files: int = 0
    params: Optional[Dict[str, Any]] = None  # Deserialized from params_json

    class Config:
        use_enum_values = True


class JobListResponse(BaseModel):
    """Response model for listing jobs."""

    jobs: List[Job]
    total: int
    limit: int
    offset: int


class JobDetailResponse(BaseModel):
    """Detailed job information including phase history."""

    id: str
    status: str
    started_at: str
    completed_at: Optional[str] = None
    project_name: str
    current_phase: Optional[Phase] = None
    tokens_used: int = 0
    total_files: int = 0
    phases: List[PhaseInfo] = []

    class Config:
        use_enum_values = True

File: tools/glass-pane/frontend/src/types/job.ts
/**
 * Job-related TypeScript types for Glass Pane frontend.
 */

export enum Phase {
  Scout = 'Scout',
  Architect = 'Architect',
  Builder = 'Builder',
  Test = 'Test',
  Screenshot = 'Screenshot',
  Documentation = 'Documentation',
  Deploy = 'Deploy',
}

export enum PhaseStatus {
  Pending = 'pending',
  Active = 'active',
  Completed = 'completed',
  Failed = 'failed',
}

export interface PhaseInfo {
  phase: Phase
  status: PhaseStatus
  description: string
  started_at: string | null
  completed_at: string | null
  iteration: number
}

export interface Job {
  id: string
  status: 'running' | 'succeeded' | 'failed' | 'cancelled'
  started_at: string // ISO 8601
  completed_at: string | null
  project_name: string
  current_phase: Phase | null
  tokens_used: number
  total_files: number
}

export interface JobDetail extends Job {
  phases: PhaseInfo[]
}

export interface JobListResponse {
  jobs: Job[]
  total: number
  limit: number
  offset: number
}

// File tree node type
export interface FileNode {
  path: string;           // Full path from root (e.g., "/src/App.tsx")
  name: string;           // File/directory name only (e.g., "App.tsx")
  type: 'file' | 'directory';
  children?: FileNode[];  // Only present for directories
  expanded: boolean;      // UI state for directory expansion
  created_at?: string;    // Optional timestamp
  size?: number;          // Optional file size in bytes
}

// Log entry type
export interface LogEntry {
  id: number;             // Numeric ID
  job_id: string;
  timestamp: string;      // ISO 8601 format
  level: LogLevel;
  message: string;
  source?: string;        // Optional: which service logged this
}

// Log level enum
export enum LogLevel {
  DEBUG = 'DEBUG',
  INFO = 'INFO',
  WARNING = 'WARNING',
  ERROR = 'ERROR'
}

// Phase update data (for SSE events)
export interface PhaseUpdateData {
  phase: string;
  status: 'starting' | 'in_progress' | 'completed' | 'failed';
  message?: string;
  timestamp: string;
}

// File created event data (for SSE events)
export interface FileCreatedData {
  path: string;
  type: 'file' | 'directory';
  timestamp: string;
}

// Token update event data (for SSE events)
export interface TokenUpdateData {
  tokens_used: number;
  total_tokens: number;
  percentage: number;
  timestamp: string;
}

// SSE event union type
export type SSEEventData =
  | { type: 'phase_update'; data: PhaseUpdateData }
  | { type: 'file_created'; data: FileCreatedData }
  | { type: 'log'; data: LogEntry }
  | { type: 'tokens'; data: TokenUpdateData };
